### PR TITLE
Upgrade base64 crate/v3

### PR DESCRIPTION
**Pull request by inashivb**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/7219

Previous PR: https://github.com/OISF/suricata/pull/11682

Changes since v3:
- Split commits as per review comment
- Rebased on top of latest master

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
  

Pipeline 22311

File: rust/src/http2/parser.rs [hunk 1]
```diff
@@ -763,7 +763,13 @@ pub fn http2_parse_frame_settings(i: &[u8]) -> IResult<&[u8], Vec<HTTP2FrameSett
 
 pub fn doh_extract_request(i: &[u8]) -> IResult<&[u8], Vec<u8>> {
     let (i, _) = tag("/dns-query?dns=")(i)?;
-    match general_purpose::STANDARD.decode(i) {
+    let mut needed_padding = 4 - (i.len() % 4);
+    let mut padded_vec = i.to_vec();
+    while needed_padding > 0 {
+        padded_vec.push(b'=');
```
Comment (catenacyber) [inline rust/src/http2/parser.rs:769]: This code should be in some util file, not specific for http2, which should just call some base64 decoder

File: rust/src/ffi/base64.rs [hunk 2]
```diff
@@ -17,6 +17,7 @@
 
 use std::os::raw::c_uchar;
 use libc::c_ulong;
+use base64::{Engine as _, engine::general_purpose};
```
Comment (catenacyber) [inline rust/src/ffi/base64.rs:20]: Why do we need `Engine as _` ?

File: rust/src/ffi/base64.rs [hunk 3]
```diff
@@ -42,7 +43,7 @@ pub unsafe extern "C" fn Base64Encode(
         return Base64ReturnCode::SC_BASE64_INVALID_ARG;
     }
     let input = std::slice::from_raw_parts(input, input_len as usize);
-    let encoded = base64::encode(input);
+    let encoded = general_purpose::STANDARD.encode(input);
```
Comment (catenacyber) [inline rust/src/ffi/base64.rs:46]: You could import final `STANDARD`

File: rust/src/http2/parser.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/http2/parser.rs:769]: Why not use `STANDARD_NO_PAD` ? cf https://docs.rs/base64/latest/base64/engine/general_purpose/constant.STANDARD_NO_PAD.html

File: rust/Cargo.lock.in [hunk 4]
```diff
@@ -3,10 +3,10 @@
 version = 3
 
 [[package]]
-name = "adler"
-version = "1.0.2"
+name = "adler2"
+version = "2.0.0"
```
Comment (catenacyber) [inline rust/Cargo.lock.in:7]: @jasonish are we forced to update everything, before updating a direct dependency ?

File: rust/src/ffi/base64.rs (hunk 3 above)
Comment (inashivb) [reply rust/src/ffi/base64.rs:46]: ok

File: rust/src/http2/parser.rs (hunk 1 above)
Comment (inashivb) [reply rust/src/http2/parser.rs:769]: Nice. Thank you. Doing in the next rev

File: rust/src/ffi/base64.rs (hunk 2 above)
Comment (inashivb) [reply rust/src/ffi/base64.rs:20]: Nothing works w/o this. I'm not sure what these are called but I've seen this before.

Comment (catenacyber) [reply rust/src/ffi/base64.rs:20]: @jasonish do you understand this rust magic ?

Comment (jasonish) [reply rust/src/ffi/base64.rs:20]: You should be able to drop the `as _` here.  `Engine` being a trait just needs to be imported, but doesn't need to be bound to a name. This syntax is usually used when binding the trait to a name would conflict with some other name in scope. I don't see a name conflict here tho.

File: rust/Cargo.lock.in (hunk 4 above)
Comment (jasonish) [reply rust/Cargo.lock.in:7]: You can get more precise if needed.

Comment (catenacyber) [reply rust/Cargo.lock.in:7]: For my information, what is the good way to proceed to update only one crate ?

Comment (jasonish) [reply rust/Cargo.lock.in:7]: > For my information, what is the good way to proceed to update only one crate ?

Never really done that.  If not updating `Cargo.toml` you can do `cargo update <package>` just to update a single package in the lock file.

If updating `Cargo.toml`, its probably a little more difficult as you'd have to figure out what transitive updates are needed or not.  So update `Cargo.toml`, build, which will regenerate `Cargo.lock`, and figure out whats needed.

**Pull request closed without merging**